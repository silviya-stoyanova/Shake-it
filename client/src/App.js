import React, { Component } from 'react';
import { BrowserRouter } from 'react-router-dom'
import { ToastContainer } from 'react-toastify'
import 'react-toastify/dist/ReactToastify.css';
import './static/css/site.css'
import './static/css/scrollbar.css'

import sessionManager from './components/utilities/session-util'
import observer from './components/utilities/observer'

import MyRoutes from './components/routes/router'
import Header from './components/common/header'
import Footer from './components/common/footer'

const defaultValue = { isLogged: false }
const { Consumer: UserInfoConsumer, Provider: UserInfoProvider } = React.createContext(defaultValue)

class App extends Component {
  constructor(props) {
    super(props)
    this.state = {
      isLogged: sessionManager.isUserLogged(),
      username: sessionManager.getUsername(),
      role: sessionManager.getUserRole()
    }

    observer.subscribe('userLogin', this.onUserLogin)
    observer.subscribe('userLogout', this.onUserLogout)
  }

  onUserLogin = () => {
    this.setState(prevState => ({ // asyncronous
      isLogged: true,
      username: sessionManager.getUsername(),
      role: sessionManager.getUserRole()
    }))
  }

  onUserLogout = () => {
    this.setState(prevState => ({
      isLogged: false,
      username: '',
      role: []
    }))
  }

  render() {
    const { isLogged, username, role } = this.state

    // console.log(this.props.navigation)

    return (
      <UserInfoProvider value={{ isLogged, username, role }}>
        <BrowserRouter>
          <ToastContainer />
          <Header />

          {
            // <div className="main-container">
          }
          <MyRoutes />
          {
            // </div>
          }

          <Footer />
        </BrowserRouter>
      </UserInfoProvider>
    )
  }

  componentDidMount() {
    sessionManager.isUserLogged()
      ? this.setState({ isLogged: true })
      : this.setState({ isLogged: false })
  }
}

export { UserInfoConsumer }
export default App