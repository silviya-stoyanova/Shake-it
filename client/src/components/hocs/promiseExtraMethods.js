import requester from '../../utilities/requests-util'
import observer from '../../utilities/observer'
import sessionManager from '../../utilities/session-util'

const promiseExtraMethods = {
    user: {
        onLoginSuccess(res) {
            sessionManager.saveSession(res.authtoken, res.username, res.role)
            observer.trigger('userLogin')
        },
        onRegisterSuccess(res, data) {
            let nextPromise = requester.login(data)
            this.handleFetchPromise(nextPromise, this.onLoginSuccess, this.onLoginFail, data) // { username: data.username, password: data.password }
        },
        onLoginFail(err) {
            this.setState(prevState => ({
                data: {
                    ...prevState.data,
                    userClass: 'correct',
                    passClass: 'error',
                    repeatPassClass: 'error'
                }
            }))
        },
        onRegisterFail(err) {
            this.setState(prevState => ({
                data: {
                    ...prevState.data,
                    userClass: 'error',
                    passClass: 'correct',
                    repeatPassClass: 'correct'
                }
            }))
        }
    },

    product: {
        onProductPromiseSuccess(err) {
            return err // do nothing
        },

        onProductPromiseFail(err) {
            if (!err.titleIsTaken) {
                this.props.history.push('/')
            }
        }
    }
}

export default promiseExtraMethods