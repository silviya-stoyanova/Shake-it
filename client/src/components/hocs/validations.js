import { toast } from 'react-toastify'

const userValidations = {
    validateUserData(formType, username, password, repeatPassword, updateState) {
        let [userClass, passClass, repeatPassClass] = ['error', 'error', 'error']
        let result = {
            username: 'invalid',
            password: 'invalid',
            repeatPassword: 'invalid'
        }

        if (username && username.trim().length >= 5) {
            userClass = 'correct'
            result.username = 'valid'
        }
        if (password && password.length >= 8) {
            passClass = 'correct'
            result.password = 'valid'
        }

        if (repeatPassword) {
            if (repeatPassword && repeatPassword.length >= 8) {
                repeatPassClass = 'correct'
                result.repeatPassword = 'valid'
            }

            if (password !== repeatPassword) {
                [passClass, repeatPassClass] = ['error', 'error']
                result.password = 'invalid'
                result.repeatPassword = 'invalid'
            }
        }

        updateState && this.setState(prevState => ({
            userInfo: {
                ...prevState.userInfo,
                userClass,
                passClass,
                repeatPassClass
            }
        }))

        return result
    },

    validateUserOnSubmit(formType, username, password, repeatPassword) {
        const isInputValid = this.validateUserData(formType, username, password, repeatPassword, false)
        let isValid = false
        let errorMsg = ''

        if (formType === 'login') {
            if (isInputValid.username === 'invalid' || isInputValid.password === 'invalid') {
                errorMsg = 'Either your username or password is invalid.'
            }
        } else if (formType === 'register') {
            if (isInputValid.password === 'invalid' || isInputValid.repeatPassword === 'invalid') {
                errorMsg = 'Both passwords must consist of at least 8 characters and they both must match!'
            }
            if (isInputValid.username === 'invalid') {
                errorMsg = 'Please enter a valid username!'
            }
        }
        if (errorMsg) {
            toast.info(errorMsg, {
                className: 'error-toast'
            })
            return isValid
        }

        isValid = true
        return isValid
    }
}

const productValidations = {
    validateProductData(formType, title, description, image, price, updateState) {
        let [titleClass, descriptionClass, priceClass] = ['error', 'error', 'error', 'error']
        let result = {
            title: 'invalid',
            description: 'invalid',
            image: 'invalid',
            price: 'invalid',
        }

        if (title && title.length > 2 && title.length <= 50) {
            titleClass = 'correct'
            result.title = 'valid'
        }
        if (description && description.length > 9 && description.length <= 250) {
            descriptionClass = 'correct'
            result.description = 'valid'
        }

        // validate only when creating a new product
        if (formType === 'create' && image) {
            result.image = 'valid'
        }
        if (price && Number(price) > 0) {
            priceClass = 'correct'
            result.price = 'valid'
        }

        updateState && this.setState(prevState => ({
            productInfo: {
                ...prevState.productInfo,
                titleClass,
                descriptionClass,
                priceClass
            }
        }))
        return result
    },

    validateProductOnSubmit(formType, title, description, image, price) {
        const updateState = false
        const isInputValid = this.validateProductData(formType, title, description, image, price, updateState)
        let isValid = false
        let errorMsg = ''

        if (isInputValid.price === 'invalid') {
            errorMsg = 'Please enter a valid price.'
        }
        // when adding a product
        if (formType === 'create' && isInputValid.image === 'invalid') {
            errorMsg = 'Please provide an image.'
        }
        if (isInputValid.description === 'invalid') {
            errorMsg = 'The description must be at least 10 symbols long.'
        }
        if (isInputValid.title === 'invalid') {
            errorMsg = 'The title length must be between 3 and 50 symbols including.'
        }
        if (errorMsg) {
            toast.info(errorMsg, {
                className: 'error-toast',
            })
            return isValid
        }

        isValid = true
        return isValid
    }
}

export { userValidations }
export { productValidations }