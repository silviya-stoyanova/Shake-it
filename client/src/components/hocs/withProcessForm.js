import React, { Component } from 'react'
import { toast } from 'react-toastify'
import requester from '../../utilities/requests-util'
import sessionManager from '../../utilities/session-util'

const withProcessForm = (Form, formType, validations, initialData, requestType, promiseExtraMethods, testServiceOnMount, testServiceOnSubmit) => {
    return class extends Component {
        constructor(props) {
            super(props)
            this.state = {
                data: initialData
            }

            this.validateData = validations ? validations.validateData.bind(this) : null
            this.validateOnSubmit = validations ? validations.validateOnSubmit.bind(this) : null

            this.promiseSuccess = promiseExtraMethods ? promiseExtraMethods.success.bind(this) : null
            this.promiseFail = promiseExtraMethods ? promiseExtraMethods.fail.bind(this) : null

            this.handleInputChange = this.handleInputChange.bind(this)
            this.handleFetchPromise = this.handleFetchPromise.bind(this)
            this.handleFormSubmit = this.handleFormSubmit.bind(this)
        }

        // static defaultProps = {
        //     service: requester.getProductInfo
        // }

        async handleInputChange({ target }) {
            console.log(target.files ? target.files[0] : null)

            await this.setState(prevState => ({
                data: {
                    ...prevState.data,
                    [target.name]: target.files ? target.files : target.value, // .trim()
                    uploadedImg: target.files ? URL.createObjectURL(target.files[0]) : prevState.data.uploadedImg
                    // uploadedImg: target.files ? target.files[0] : prevState.data.uploadedImg
                }
            }))

            const updateState = true
            const { data } = this.state

            // console.log(data.uploadedImg)

            this.validateData(formType, data, updateState)
        }

        handleFetchPromise(promise, onSuccessFunc, onFailFunc, data = '') {
            promise.then(res => {
                if (!res.ok) {
                    return Promise.reject(res)
                }
                return res.json()

            }).then(res => {
                toast.info(res.success, {
                    className: 'success-toast'
                })

                // data.redirectOnSuccess &&
                this.props.history.push('/')

                onSuccessFunc && onSuccessFunc(res, data)

            }).catch(err => {
                err.json().then(error => {
                    toast.info(error.message, {
                        className: 'error-toast'
                    })

                    onFailFunc && onFailFunc(error)
                })
            })
        }

        async handleFormSubmit(event) {
            // console.log(event.preventDefault())
            
            event.preventDefault()
            const { data } = this.state
            const jwtToken = sessionManager.getUserInfo().authtoken
            let isValid = this.validateOnSubmit(formType, data)

            if (!isValid) {
                return
            }

            if (testServiceOnSubmit) {
                testServiceOnSubmit()
                    // to see the message inside the response, uncomment the next two commented out lines
                    // .then(res => res.json())
                    .then(res => {
                        // console.log(res.success)
                        this.props.history.push('/')
                    })
                    .catch(err => {
                        return err
                    })

            } else {
                const promise = requester[requestType](data, jwtToken)
                this.handleFetchPromise(promise, this.promiseSuccess, this.promiseFail, data)
            }
        }

        render() {
            return <Form {...this.state} handleInputChange={this.handleInputChange} handleFormSubmit={this.handleFormSubmit} />
        }

        componentDidMount() {
            if (formType === 'create' || formType === 'login' || formType === 'register') {
                return
            }
            // let promise = requester.getProductInfo(this.state.productInfo._id)
            // this.handleFetchPromise(promise, null, null, { redirectOnFail: true })
            // requester.getProductInfo(this.state.data._id)

            const serviceReq = testServiceOnMount || requester.getProductInfo  // new

            const productId = this.props.match                      // new
                ? this.props.match.params.productId
                : null
            // const productId = this.props.match.params.productId

            // requester.getProductInfo(productId)                  // prev
            serviceReq && serviceReq(productId)                     // new
                .then(res => {
                    if (!res.ok) {
                        return Promise.reject(res)
                    }
                    return res.json()
                })
                .then(async res => {
                    await this.setState({
                        data: res
                    })
                })
                .catch(error => {
                    this.props.history.push('/')

                    error.json().then(err => {
                        toast.info(err.message, {
                            className: 'error-toast'
                        })
                    })
                })
        }
    }
}

export default withProcessForm