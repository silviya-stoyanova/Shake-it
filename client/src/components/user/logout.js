import React from 'react'
import { Redirect } from 'react-router-dom'
import { toast } from 'react-toastify'
import sessionManager from '../utilities/session-util'
import observer from '../utilities/observer'

function logout() {
    try {
        sessionManager.clearSession()
        observer.trigger('userLogout')
        toast.info('Successful log out! üçπ', {
            className: 'success-toast',
        })

    } catch (err) {
        toast.info('Something went wrong when trying to log you out..', {
            className: 'error-toast',
        })
        console.log(err)
    }

    return <Redirect to='/' />
}

export default logout