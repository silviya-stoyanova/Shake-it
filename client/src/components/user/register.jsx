import React, { Component } from 'react'
import { Redirect } from 'react-router-dom'
import { toast } from 'react-toastify'
import '../../static/css/forms.css'

import { UserInfoConsumer } from '../../App'
import observer from '../utilities/observer'
import requester from '../utilities/requests-util'
import sessionManager from '../utilities/session-util'

class UserRegister extends Component {
    constructor(props) {
        super(props)
        this.state = {
            username: '',
            password: '',
            repeatPassword: '',

            userClass: '',
            passClass: '',
            repeatPassClass: '',
        }

        this.handleInputChange = this.handleInputChange.bind(this)
    }

    validateInputFields = (username, password, repeatPassword) => {
        let [userClass, passClass, repeatPassClass] = ['error', 'error', 'error']

        const result = {
            username: 'invalid',
            password: 'invalid',
            repeatPassword: 'invalid',
        }

        if (username && username.trim().length) {
            userClass = 'correct'
            result.username = 'valid'
        }
        if (password && password.trim().length) {
            passClass = 'correct'
            result.password = 'valid'
        }
        if (repeatPassword && repeatPassword.trim().length) {
            repeatPassClass = 'correct'
            result.repeatPassword = 'valid'
        }

        if (password !== repeatPassword) {
            [passClass, repeatPassClass] = ['error', 'error']
            result.password = 'invalid'
            result.repeatPassword = 'invalid'
        }

        this.setState({ userClass, passClass, repeatPassClass })
        return result
    }

    async handleInputChange({ target }) {
        await this.setState({
            [target.name]: target.value
        })

        const { username, password, repeatPassword } = this.state
        this.validateInputFields(username, password, repeatPassword)
    }

    handleFormSubmit = async (event) => {
        event.preventDefault()
        const { username, password, repeatPassword } = this.state
        const isInputValid = this.validateInputFields(username, password, repeatPassword)

        if (isInputValid.username === 'invalid') {
            return toast.info('Please enter a valid username!', {
                className: 'error-toast'
            })
        }
        if (isInputValid.password === 'invalid' &&
            isInputValid.repeatPassword === 'invalid') {
            return toast.info('Both passwords must match!', {
                className: 'error-toast'
            })
        }

        requester.register(username, password)
            .then(res => {
                if (!res.ok) {
                    return Promise.reject(res)
                }
                return res.json()
            })
            .then(res => {

                toast.info('Successful registration! 🍹', {
                    className: 'success-toast',
                })

                requester.login(username, password)
                    .then(res => {
                        if (!res.ok) {
                            return Promise.reject(res)
                        }
                        return res.json()
                    })
                    .then(res => {
                        sessionManager.saveSession(res.authtoken, res.username, res.role)
                        observer.trigger('userLogin')
                    })
                    .catch(err => {
                        err.json().then(error => {
                            toast.info(error.message, {
                                className: 'error-toast',
                            })
                        })
                        this.setState({
                            userClass: 'error'
                        })
                    })
            })
            .catch(err => {
                console.log(err)

                err.json().then(error => {
                    toast.info(error.message, {
                        className: 'error-toast',
                    })
                })
                this.setState({
                    userClass: 'error'
                })
            })
    }

    render() {
        const { userClass, passClass, repeatPassClass } = this.state

        return (
            <UserInfoConsumer>
                {data =>
                    <div className="form">
                        {data.isLogged
                            ? <Redirect to='/' />
                            : null
                        }
                        <form onSubmit={this.handleFormSubmit}>
                            <div className="form-type">Register</div>
                            <hr />

                            <div className="form-fields-wrapper">
                                <label htmlFor="username">Username</label>
                                <input autoFocus onChange={this.handleInputChange} name="username" className={userClass} type="text" id="username" />

                                <label htmlFor="password">Password</label>
                                <input onChange={this.handleInputChange} name="password" className={passClass} type="password" id="password" />

                                <label htmlFor="repeat-password">Repeat password</label>
                                <input onChange={this.handleInputChange} name="repeatPassword" className={repeatPassClass} type="password" id="repeat-password" />
                            </div>

                            <button className="button" type="submit">Submit</button>
                        </form>
                    </div>
                }
            </UserInfoConsumer>
        )
    }
}

export default UserRegister