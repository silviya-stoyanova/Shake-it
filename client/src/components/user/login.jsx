import React from 'react'
import '../../static/css/forms.css'
import '../../static/css/notifications.css'
import useTitle from '../page-title/useTitle'
import withProcessForm from '../hocs/withProcessForm'
import { userValidations } from '../hocs/validations'
import promiseExtraMethods from '../hocs/promiseExtraMethods'

const UserLogin = (props) => {
    useTitle('Login')
    const { handleInputChange, handleFormSubmit } = props
    const { userClass, passClass } = props.data
    // const { userClass, passClass } = props.userInfo

    return (
        <React.Fragment>
            <div className="form">
                <form id="login-form" onSubmit={handleFormSubmit} >
                    <div className="form-type">Login</div>
                    <hr />

                    <div className="form-fields-wrapper">
                        <label htmlFor="username">Username</label>
                        <input autoFocus onChange={handleInputChange} name="username" className={userClass} type="text" id="username" />

                        <label htmlFor="password">Password</label>
                        <input onChange={handleInputChange} name="password" className={passClass} type="password" id="password" />
                    </div>

                    <button className="button" type="submit">Submit</button>
                </form>
            </div>
        </React.Fragment>
    )
}

const initialData = {
    username: '',
    password: '',

    userClass: '',
    passClass: '',
}

const requestType = 'login'

const extraMethods = {
    success: promiseExtraMethods.user.onLoginSuccess,
    fail: promiseExtraMethods.user.onLoginFail
}

//* wrong: export default <withProcessForm Form={UserLogin} />
export default withProcessForm(UserLogin, 'login', userValidations, initialData, requestType, extraMethods)