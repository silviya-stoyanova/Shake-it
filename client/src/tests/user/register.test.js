import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import { mount } from 'enzyme'
import AuthRoute from '../../components/routes/auth-route'
import withProcessForm from '../../components/hocs/withProcessForm'
import { userValidations } from '../../components/hocs/validations'
import promiseExtraMethods from '../../components/hocs/promiseExtraMethods'
import { UserRegister } from '../../components/user/register'
import { UserInfoProvider } from '../../App'


// scenarious:

// logged in user tries to reach the Register page  ==> should be redirected
// logged in admin tries to reach the Register page  ==> should be redirected
// logged out user tries to reach the Register page ==> should be able to view Register page

// -- submit button is not clicked yet
//    -- valid info is passed to all fields
//    -- invalid info is passed for [username] / [password] / [repeatPassword]
//    -- invalid info is passed for [password] and [repeat-password]
//    -- invalid info is passed for [username, password and repeatPassword]
//
// -- submit button is clicked
//    -- valid info is passed to to all fields                         *   sampleEvents.onFormSubmit should BE Called !!   *
//    -- invalid info is passed for [username] / [password] / [rePass] * sampleEvents.onFormSubmit should NOT be Called !! *
//    -- invalid info is passed for [password] and [repeat-password]   * sampleEvents.onFormSubmit should NOT be Called !! *
//    -- invalid info is passed for [username, password and rePass]    * sampleEvents.onFormSubmit should NOT be Called !! *


const initialData = {
    username: '',
    password: '',
    repeatPassword: '',

    userClass: '',
    passClass: '',
    repeatPassClass: ''
}

const extraMethods = {
    success: promiseExtraMethods.user().onRegisterSuccess,
    fail: promiseExtraMethods.user().onRegisterFail
}

const mockRegisterSubmit = () => {
    const sampleReturnedData = JSON.stringify({ success: 'Successful registration! â™¥' })
    return Promise.resolve(new Response(sampleReturnedData))
}

const sampleEvents = {
    username: {
        validUsername: {
            target: { name: 'username', value: 'cherry' }
        },
        emptyUsername: {
            target: { name: 'username', value: '' }
        },
        tooShortUsername: {
            target: { name: 'username', value: 'yam' }
        }
    },
    password: {
        validPassword: {
            target: { name: 'password', value: 'pass-licious' }
        },
        emptyPassword: {
            target: { name: 'password', value: '' }
        },
        tooShortPassword: {
            target: { name: 'password', value: 'pass' }
        }
    },
    repeatPassword: {
        validRepeatPassword: {
            target: { name: 'repeatPassword', value: 'pass-licious' }
        },
        // long enough but not the same as validPassword
        diffRepeatPassword: {
            target: { name: 'repeatPassword', value: 'delicious' }
        },
        emptyRepeatPassword: {
            target: { name: 'repeatPassword', value: '' }
        },
        tooShortRepeatPassword: {
            target: { name: 'repeatPassword', value: 'pass' }
        }
    },
    onFormSubmit: {
        preventDefault: () => {
            return 'I am fake prevent-defaulter.'
        }
    }
}

const historyMock = {
    push: jest.fn()
}

describe('tests for the component UserRegister', () => {
    let TestHoc = withProcessForm(UserRegister, 'register', userValidations, initialData, null, extraMethods, null, mockRegisterSubmit)
    let wrapper

    beforeEach(() => {
        wrapper = mount(
            <TestHoc history={historyMock} />
        )
    })

    afterEach(() => {
        // to reset Jest mock function calls count after every test:
        historyMock.push.mockClear()
    })

    describe('guest functionalities', () => {

        it('should allow [guests] to view the register page', () => {
            const defaultUserValue = { isLogged: false, username: '', role: '' }

            wrapper = mount(
                <BrowserRouter>
                    <UserInfoProvider value={defaultUserValue} >
                        <AuthRoute path="/user/register" component={UserRegister} role="null" />
                    </UserInfoProvider>
                </BrowserRouter>
            )

            expect(wrapper.find('Route[path="/user/register"]').length).toEqual(1)
        })

        it('should render the register form with empty fields', async () => {
            expect(wrapper.html()).toMatchSnapshot()

            await wrapper.find('form').simulate('submit', sampleEvents.onFormSubmit)
            expect(historyMock.push).not.toBeCalled()
            expect(historyMock.push.mock.calls.length).toEqual(0)
        })

        it('should mark as [valid] all fields of the form', async () => {
            await wrapper.find('input[id="username"]').simulate('change', sampleEvents.username.validUsername)
            expect(wrapper.find('input[id="username"]').prop('value')).toEqual(sampleEvents.username.validUsername.target.value)
            wrapper.update()
            expect(wrapper.find('input[id="username"]').hasClass('correct')).toEqual(true)

            await wrapper.find('input[id="password"]').simulate('change', sampleEvents.password.validPassword)
            expect(wrapper.find('input[id="password"]').prop('value')).toEqual(sampleEvents.password.validPassword.target.value)
            wrapper.update()
            expect(wrapper.find('input[id="password"]').hasClass('correct')).toEqual(true)

            await wrapper.find('input[id="repeat-password"]').simulate('change', sampleEvents.repeatPassword.validRepeatPassword)
            expect(wrapper.find('input[id="repeat-password"]').prop('value')).toEqual(sampleEvents.repeatPassword.validRepeatPassword.target.value)
            wrapper.update()
            expect(wrapper.find('input[id="repeat-password"]').hasClass('correct')).toEqual(true)

            await wrapper.find('form').simulate('submit', sampleEvents.onFormSubmit)
            expect(historyMock.push).toBeCalled()
            expect(historyMock.push).toHaveBeenLastCalledWith('/')
            expect(historyMock.push.mock.calls.length).toEqual(1)
        })

        it('should mark as invalid [username] after change', async () => {
            await wrapper.find('input[id="username"]').simulate('change', sampleEvents.username.emptyUsername)
            expect(wrapper.find('input[id="username"]').prop('value')).toEqual(sampleEvents.username.emptyUsername.target.value)
            wrapper.update()
            expect(wrapper.find('input[id="username"]').hasClass('error')).toEqual(true)

            await wrapper.find('input[id="username"]').simulate('change', sampleEvents.username.tooShortUsername)
            expect(wrapper.find('input[id="username"]').prop('value')).toEqual(sampleEvents.username.tooShortUsername.target.value)
            wrapper.update()
            expect(wrapper.find('input[id="username"]').hasClass('error')).toEqual(true)

            await wrapper.find('form').simulate('submit', sampleEvents.onFormSubmit)
            expect(historyMock.push).not.toBeCalled()
            expect(historyMock.push.mock.calls.length).toEqual(0)
        })

        it('should mark as invalid [password] after change', async () => {
            await wrapper.find('input[id="password"]').simulate('change', sampleEvents.password.emptyPassword)
            expect(wrapper.find('input[id="password"]').prop('value')).toEqual(sampleEvents.password.emptyPassword.target.value)
            wrapper.update()
            expect(wrapper.find('input[id="password"]').hasClass('error')).toEqual(true)

            await wrapper.find('input[id="password"]').simulate('change', sampleEvents.password.tooShortPassword)
            expect(wrapper.find('input[id="password"]').prop('value')).toEqual(sampleEvents.password.tooShortPassword.target.value)
            wrapper.update()
            expect(wrapper.find('input[id="password"]').hasClass('error')).toEqual(true)

            await wrapper.find('form').simulate('submit', sampleEvents.onFormSubmit)
            expect(historyMock.push).not.toBeCalled()
            expect(historyMock.push.mock.calls.length).toEqual(0)
        })

        it('should mark as invalid [repeat-password] after change', async () => {
            await wrapper.find('input[id="repeat-password"]').simulate('change', sampleEvents.repeatPassword.emptyRepeatPassword)
            expect(wrapper.find('input[id="repeat-password"]').prop('value')).toEqual(sampleEvents.repeatPassword.emptyRepeatPassword.target.value)
            wrapper.update()
            expect(wrapper.find('input[id="repeat-password"]').hasClass('error')).toEqual(true)

            await wrapper.find('input[id="repeat-password"]').simulate('change', sampleEvents.repeatPassword.tooShortRepeatPassword)
            expect(wrapper.find('input[id="repeat-password"]').prop('value')).toEqual(sampleEvents.repeatPassword.tooShortRepeatPassword.target.value)
            wrapper.update()
            expect(wrapper.find('input[id="repeat-password"]').hasClass('error')).toEqual(true)

            await wrapper.find('form').simulate('submit', sampleEvents.onFormSubmit)
            expect(historyMock.push).not.toBeCalled()
            expect(historyMock.push.mock.calls.length).toEqual(0)
        })

        it('should mark as invalid both [password] and [repeat-password] after change', async () => {
            await wrapper.find('input[id="password"]').simulate('change', sampleEvents.password.validPassword)
            expect(wrapper.find('input[id="password"]').prop('value')).toEqual(sampleEvents.password.validPassword.target.value)

            await wrapper.find('input[id="repeat-password"]').simulate('change', sampleEvents.repeatPassword.diffRepeatPassword)
            expect(wrapper.find('input[id="repeat-password"]').prop('value')).toEqual(sampleEvents.repeatPassword.diffRepeatPassword.target.value)

            wrapper.update()

            expect(wrapper.find('input[id="password"]').hasClass('error')).toEqual(true)
            expect(wrapper.find('input[id="repeat-password"]').hasClass('error')).toEqual(true)

            await wrapper.find('form').simulate('submit', sampleEvents.onFormSubmit)
            expect(historyMock.push).not.toBeCalled()
            expect(historyMock.push.mock.calls.length).toEqual(0)
        })

        it('should mark as [invalid] all fields of the form', async () => {
            await wrapper.find('input[id="username"]').simulate('change', sampleEvents.username.tooShortUsername)
            await wrapper.find('input[id="password"]').simulate('change', sampleEvents.password.tooShortPassword)
            await wrapper.find('input[id="repeat-password"]').simulate('change', sampleEvents.repeatPassword.tooShortRepeatPassword)

            expect(wrapper.find('input[id="username"]').prop('value')).toEqual(sampleEvents.username.tooShortUsername.target.value)
            expect(wrapper.find('input[id="password"]').prop('value')).toEqual(sampleEvents.password.tooShortPassword.target.value)
            expect(wrapper.find('input[id="repeat-password"]').prop('value')).toEqual(sampleEvents.repeatPassword.tooShortRepeatPassword.target.value)

            wrapper.update()

            expect(wrapper.find('input[id="username"]').hasClass('error')).toEqual(true)
            expect(wrapper.find('input[id="password"]').hasClass('error')).toEqual(true)
            expect(wrapper.find('input[id="repeat-password"]').hasClass('error')).toEqual(true)

            await wrapper.find('form').simulate('submit', sampleEvents.onFormSubmit)
            expect(historyMock.push).not.toBeCalled()
            expect(historyMock.push.mock.calls.length).toEqual(0)
        })
    })

    describe('user functionalities', () => {
        const defaultUserValue = { isLogged: true, username: 'cupcake', role: 'User' }

        it('should redirect [users] to the home page when they accesses register page', () => {
            wrapper = mount(
                <BrowserRouter>
                    <UserInfoProvider value={defaultUserValue} >
                        <AuthRoute path="/user/register" component={UserRegister} role="null" />
                    </UserInfoProvider>
                </BrowserRouter>
            )

            expect(wrapper.find('Redirect[to="/"]').length).toEqual(1)
        })

    })

    describe('admin functionalities', () => {
        const defaultUserValue = { isLogged: true, username: 'mushroom4e', role: 'Admin' }

        it('should let The [Admin] view the register page', () => {
            wrapper = mount(
                <BrowserRouter>
                    <UserInfoProvider value={defaultUserValue} >
                        <AuthRoute path="/user/register" component={UserRegister} role="null" />
                    </UserInfoProvider>
                </BrowserRouter>
            )

            expect(wrapper.find('Route[path="/user/register"]').length).toEqual(1)
        })

    })
})