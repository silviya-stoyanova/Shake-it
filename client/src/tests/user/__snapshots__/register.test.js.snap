// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for the component UserRegister guest functionalities should render the register form with empty fields 1`] = `"<div class=\\"form\\"><form><div class=\\"form-type\\">Register</div><hr><div class=\\"form-fields-wrapper\\"><label for=\\"username\\">Username</label><input name=\\"username\\" class=\\"\\" type=\\"text\\" id=\\"username\\" value=\\"\\"><label for=\\"password\\">Password</label><input name=\\"password\\" class=\\"\\" type=\\"password\\" id=\\"password\\" value=\\"\\"><label for=\\"repeat-password\\">Repeat password</label><input name=\\"repeatPassword\\" class=\\"\\" type=\\"password\\" id=\\"repeat-password\\" value=\\"\\"></div><button class=\\"button\\" type=\\"submit\\">Submit</button></form></div>"`;
